src/arfsdao.ts

```diff
@@ -750,74 +768,387 @@ export class ArFSPrivateDrive extends ArFSEntity implements ArFSDriveEntity {
 		super(appName, appVersion, arFS, contentType, driveId, entityType, name, 0, txId, unixTime);
 	}
 }
-export class ArFSPrivateDriveBuilder {
-	appName?: string;
-	appVersion?: string;
-	arFS?: string;
-	contentType?: ContentType;
-	driveId?: DriveID;
-	entityType?: EntityType;
-	name?: string;
-	txId?: TransactionID;
-	unixTime?: number;
-	drivePrivacy?: DrivePrivacy;
-	rootFolderId?: FolderID;
-	driveAuthMode?: string;
-	cipher?: string;
-	cipherIV?: string;
-
-	async build(wallet: JWKWallet, drivePassword: string, arweave: Arweave): Promise<ArFSPrivateDrive> {
-		if (
-			this.appName?.length &&
-			this.appVersion?.length &&
-			this.arFS?.length &&
-			this.contentType?.length &&
-			this.driveId?.length &&
-			this.entityType?.length &&
-			this.txId?.length &&
-			this.unixTime &&
-			this.drivePrivacy?.length &&
-			this.driveAuthMode?.length &&
-			this.cipher?.length &&
-			this.cipherIV?.length
-		) {
-			const txData = await arweave.transactions.getData(this.txId, { decode: true });
-			const dataBuffer = Buffer.from(txData);
-			const driveKey: Buffer = await deriveDriveKey(
-				drivePassword,
-				this.driveId,
-				JSON.stringify(wallet.getPrivateKey())
-			);
 
-			const decryptedDriveBuffer: Buffer = await driveDecrypt(this.cipherIV, driveKey, dataBuffer);
-			const decryptedDriveString: string = await Utf8ArrayToStr(decryptedDriveBuffer);
-			const decryptedDriveJSON = await JSON.parse(decryptedDriveString);
+export class ArFSFileOrFolderEntity extends ArFSEntity implements ArFSFileFolderEntity {
+	lastModifiedDate!: never;
+	folderId?: string;
+
+	constructor(
+		appName: string,
+		appVersion: string,
+		arFS: string,
+		contentType: string,
+		driveId: string,
+		entityType: string,
+		name: string,
+		txId: string,
+		unixTime: number,
+		readonly parentFolderId: string,
+		readonly entityId: string
+	) {
+		super(appName, appVersion, arFS, contentType, driveId, entityType, name, 0, txId, unixTime);
+	}
+}
+
+export interface ArFSWithPath {
+	readonly path: string;
+	readonly txIdPath: string;
+	readonly entityIdPath: string;
+}
+
+export class ArFSPublicFileOrFolderData extends ArFSFileOrFolderEntity implements ArFSWithPath {
+	constructor(
+		appName: string,
+		appVersion: string,
+		arFS: string,
+		contentType: string,
+		driveId: string,
+		entityType: string,
+		name: string,
+		txId: string,
+		unixTime: number,
+		parentFolderId: string,
+		readonly entityId: string,
+		readonly path: string,
+		readonly txIdPath: string,
+		readonly entityIdPath: string
+	) {
+		super(
+			appName,
+			appVersion,
+			arFS,
+			contentType,
+			driveId,
+			entityType,
+			name,
+			txId,
+			unixTime,
+			parentFolderId,
+			entityId
+		);
+	}
+}
+
+export class ArFSPrivateFileOrFolderData extends ArFSFileOrFolderEntity implements ArFSWithPath {
```

Comment:
How about a generic type to save on some code?
```
export class EntityWithPaths<T extends ArFSFileOrFolderEntity> {
	constructor(readonly entity: T, readonly path: string, readonly txIdPath: string, readonly entityIdPath: string) {}
}
```
Use it like:
```
const mergedData = allChildrenOfFolder.map((entity) => {
			const path = `${hierarchy.pathToFolderId(entity.parentFolderId)}${entity.name}`;
			const txPath = `${hierarchy.txPathToFolderId(entity.parentFolderId)}${entity.txId}`;
			const entityIdPath = `${hierarchy.entityPathToFolderId(entity.parentFolderId)}${entity.entityId}`;
			return new EntityWithPaths<ArFSFileOrFolderEntity>(entity, path, txPath, entityIdPath);
		});
```
Only caveat is you'll need a `toJSON` function to format the output how you'd like at the commander layer.

Reply:
Quick and dirty implementation of `toJson` could be something that removes the trailing `}` of the entity JSON and then creates a format string like:
```
return `
${clippedEntityJSON},
    path: "${this.path}",
    txPath: "${this.txPath}",
    entityPath: "${this.entityPath}",
}
`;
```
I need a bath after this...

Reply:
I found a way to apply this to the classes we were already using: a7d657d715e552dc6905c61eaa0593fc759166e3

Reply:
That commit introduced a bug, fixed at: fa8ee61
`JSON.stringify()` was trying to parse the tree instance, which is a circular structure